pkg/cmd/checkendpoints/controller/backoff_recorder.go

```diff
@@ -0,0 +1,222 @@
+	}
+}
+
+func WithLongWindow(duration time.Duration, maxCount int) backoffEventRecorderOptionFunc {
+	return func(recorder *backoffEventRecorder) {
+		if recorder.longWindowTicker != nil {
+			recorder.longWindowTicker.Stop()
+		}
+		recorder.longWindow = duration
+		recorder.longWindowTicker = time.NewTicker(duration)
+		recorder.longWindowCountMax = maxCount
+	}
+}
+
+func WithBackoff(duration time.Duration) backoffEventRecorderOptionFunc {
+	return func(recorder *backoffEventRecorder) {
+		if recorder.backoffTicker != nil {
+			recorder.backoffTicker.Stop()
+		}
+		recorder.backoff = duration
+		recorder.backoffTicker = nil
+	}
+}
+
+func (r *backoffEventRecorder) Event(reason, message string) {
+	r.event(corev1.EventTypeNormal, reason, message)
+}
+
+func (r *backoffEventRecorder) Eventf(reason, messageFmt string, args ...interface{}) {
+	r.Event(reason, fmt.Sprintf(messageFmt, args...))
+}
+
+func (r *backoffEventRecorder) Warning(reason, message string) {
+	r.event(corev1.EventTypeWarning, reason, message)
+}
+
+func (r *backoffEventRecorder) Warningf(reason, messageFmt string, args ...interface{}) {
+	r.Warning(reason, fmt.Sprintf(messageFmt, args...))
+}
+
+type backoffEventRecorder struct {
+	// the wrapped event recorder
+	recorder events.Recorder
+
+	// lock must be held to update any internal state
+	lock sync.Mutex
+
+	// events waiting to be emitted
+	events map[string]map[string][]eventInfo
+
+	// short excessive event window
+	shortWindow         time.Duration
+	shortWindowCountMax int
+
+	// excessive event long window
+	longWindow         time.Duration
+	longWindowCountMax int
+
+	// backoff delay when excessive events detected
+	backoff time.Duration
+
+	// keep count of events during windows
+	shortWindowCount int
+	longWindowCount  int
+
+	// tickers
+	shortWindowTicker *time.Ticker
+	longWindowTicker  *time.Ticker
+
+	// backoffTicker is nil if no backoff delay is in progress
+	backoffTicker *time.Ticker
+}
+
+type eventInfo struct {
+	timestamp time.Time
+	message   string
+}
+
+func (e eventInfo) String() string {
+	return fmt.Sprintf("%s: %s", e.timestamp.Format(time.RFC3339), e.message)
+}
+
+func (r *backoffEventRecorder) event(eventType, reason, message string) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// add event to buffer
+	if r.events == nil {
+		r.events = make(map[string]map[string][]eventInfo)
+	}
+	if _, ok := r.events[eventType]; !ok {
+		r.events[eventType] = map[string][]eventInfo{}
+	}
+	r.events[eventType][reason] = append(r.events[eventType][reason], eventInfo{
+		timestamp: time.Now(),
+		message:   message,
+	})
+
+	// handle backoff period
+	if r.backoffTicker != nil {
+		select {
+		case <-r.backoffTicker.C:
+			// backoff period has passed
+			klog.V(1).Info("Resuming connectivity event emission.")
+			r.backoffTicker = nil
+			r.longWindowCount = 0
+			r.shortWindowCount = 0
+		default:
+			// still in backoff period
+			return
+		}
+	}
+
+	// update window event count
```

Comment:
seeing this logic, wondering whether we already have something similar? Maybe in client-go or apimachinery?

Reply:
I could not find something equivalent.